Add blog post on integration testing in Kyma (#526)
kyma-project/website: content/blog-posts/2020-01-10-integration-testing-in-k8s/index.md

---

> title: "How we approached integration testing in Kubernetes, and why we stopped using Helm tests"

Reviewer: I would prefer `Our approach to integration testing in Kubernetes, and why we stopped using Helm tests` but will not argue for it :)

Author: This way it is simply consistent as both sentences start with question words.

author:
  name: "Adam Szecowka and Karolina Zydek"
tags:
  - integration-testing
  - octopus
  - helm-tests
redirectFrom:
  - "/blog/2020-01-09-integration-testing-in-k8s"
---

In Kubernetes, you often come across projects that are true mosaics of cloud-native applications. In such a microservice architecture, we don't meet too many free-standing services - most of them have dependencies that aren't immediately obvious. Integrating such pieces and checking if they all work together may be a daunting challenge.

<!-- overview -->

> The structure of Kubernetes projects usually consists of a number of Helm charts that you can roughly divide into these categories:

Author: As agreed, I'll leave it as it is.

## Helm tests and related issues

When we started working on Kyma, we had all those things in mind. We decided to define integration tests as [**Helm tests**](https://helm.sh/docs/topics/chart_tests/). In this approach, a test is a [Kubernetes Job](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/) with the `helm.sh/hook: test` annotation. You place the test under the `templates` directory of the given Helm chart. Helm creates such tests in a Kubernetes cluster just like it does with other resources.

The reason why we took this testing path was quite simple - we used Helm extensively in our project and the Helm in-built tool for testing was a natural choice. Also, writing Helm tests turned out to be quite easy.

As our project grew, however, we came across a few obstacles that painfully hindered our work, and that we couldn't easily address with Helm tests at that time:

- Running the whole suite of integration tests took ages so we needed an easy way of selecting tests we want to run.

> - The number of flaky tests increased, and we wanted to ensure they are automatically retried.

Author: I'll change it.

- We needed a way of verifying the tests' stability and detecting flaky tests.
- We wanted to run tests concurrently to reduce the overall testing time.

At that point, we decided we need a more powerful tool. Since we couldn't find a project that would serve all our needs, we developed one on our own.

## The rise of Octopus

> This is how [**Octopus**](https://github.com/kyma-incubator/octopus/blob/master/README.md) was born. In short, Octopus is a Kubernetes controller which operates on two customer resources called [**TestDefinition**](https://github.com/kyma-incubator/octopus/blob/master/docs/crd-test-definition.md) and [**ClusterTestSuite**](https://github.com/kyma-incubator/octopus/blob/master/docs/crd-cluster-test-suite.md).

Author: There is no particular rule behind it - I was just thinking that emphasizing the new and important term would be a good idea. I can remove it as we already have links in those terms so they are already quite visible.

```
---
apiVersion: testing.kyma-project.io/v1alpha1
kind: TestDefinition
metadata:
  labels:
    component/service-catalog: "true" # You can define your own labels here
  name: test-example
spec:
  template:
    spec:
      containers:
        - name: test
          image: alpine:latest
          command:
            - "pwd"
```

ClusterTestSuite, on the other hand, defines which tests to run on a cluster and how to do that. In the below example, we define that we want to run all tests with the `component=service-catalog` label and we specify that every test should be executed once (`count=1`). If a test from the suite fails, we will retry it once (`maxRetries=1`). Moreover, tests can be executed concurrently, in the maximum number of two at a time (`concurrency=2`).

```
apiVersion: testing.kyma-project.io/v1alpha1
kind: ClusterTestSuite
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: testsuite-selected-by-labels
spec:
  count: 1
  maxRetries: 1
  concurrency: 2
  selectors:
    matchLabelExpressions:

>       - component=service-catalog

Reviewer: TestDefinition has
```
component/service-catalog: "true"
```

are you sure this is correct then?
```
matchLabelExpressions:
      - component=service-catalog
```

shouldn't testdefinition have label:
```
component:service-catalog
```

Author: It turned out you were right - I corrected the label.

```

When Octopus notices a new ClusterTestSuite, it calculates which tests should be executed, and then schedules them according to the test suite specification. All information about the status of test execution is stored in the **status** field. This status informs you which tests will be executed, which of them were already executed, which succeeded and which failed, which tests were retried, and how much time all tests and every single one took. See this example:

```
apiVersion: testing.kyma-project.io/v1alpha1
kind: ClusterTestSuite
metadata:
  name: testsuite-all
status:
  completionTime: 2019-03-06T12:37:20Z
  conditions:
  - status: "True"
    type: Succeeded
  results:
  - executions:
    - completionTime: 2019-03-06T12:37:20Z
      id: octopus-testing-pod-jz9qq
      podPhase: Succeeded
      startTime: 2019-03-06T12:37:15Z
    name: test-example
    namespace: default
    status: Succeeded
  startTime: 2019-03-06T12:37:15Z
  ```

With Octopus, all test preparation steps come down to creating:

1. Test in the language of your choice (yes, Octopus is language-agnostic).
2. TestDefinition which specifies the image to use and commands to run.

> 3. ClusterTestSuite that specifies which tests to run on a cluster, and how you want to run them.

Author: I put that in this point and the previous one o make it consistent, but not to repeat `which` in one sentence.

3. **Running tests multiple times**

   You can easily specify in a ClusterTestSuite how many times every test should be executed by adding the **count** field. That can be particularly useful to detect flaky tests or ensure that a newly created test is stable.

4. **Full support for concurrent testing**

   You can define in the ClusterTestSuite how many tests can be executed at the same time by adding the **concurrency** field. In our integration Prow jobs, we define the ClusterTestSuite with **concurrency** set to `5`. All tests are executed in around 20 minutes, but if they were executed one after another, they would take twice as long. Thanks to concurrency, we are saving time and money, developers have immediate feedback, and clusters created for executing tests are removed faster.

>    You can also specify on the TestDefinition level if you want to execute the given test from running concurrently as part of the ClusterTestSuite (**disableConcurrency**). That feature might be useful in cases when you don't want to run a test with dependencies on other tests from the suite.

Author: True 👍

   In general, the concurrency level which you define depends on the size of your cluster. Every Pod consumes some amount of CPU and memory, and we need to take those two parameters into account when defining the concurrency level.

5. **Visibility**

   Octopus gave us much more insight into test definitions and results than we had with Helm tests. After executing the ClusterTestSuite, you can easily analyze how much time each test from the suite takes and identify the problematic ones.

   - **CLI** - We integrated Octopus with Kyma Command Line Interface (CLI). This means that you can use simple commands to get test definitions (`kyma test definitions`), run tests with selected flags (`kyma test run --concurrency=5 --max-retries=1`), or watch tests execution status (`watch kyma test status`).

>     [Here](https://asciinema.org/a/287696) you can take a look at Kyma CLI in action.

Reviewer 2: This line should go under the bullet point, I don't know why but when rendered it is moved to the left.

Author: Hmm, it seems to look fine:

![image](https://user-images.githubusercontent.com/40655785/72525838-58269f80-3865-11ea-876c-32124b81046e.png)

   - **Dashboards** - We used information available in the **status** field of the ClusterTestSuite to visualize test details on Prow dashboards. In the below example, you can clearly see all details of the `post-master-kyma-gke-integration` Prow job that builds our artifacts on a GKE cluster after every merge to the `master` branch.

>    ![Testing time](./test-status.png)

Reviewer 2: Same situation as with the CLI link - should be aligned with the `Dashboards` point.

Author: I checked and it won't have an indent on the blog section :(

- Define the maximum duration of the ClusterTestSuite after which test executions are interrupted and marked as failed. (**suiteTimeout**)
- Indicate that a test shouldn't be executed. (**skip**)
- Specify the maximum duration of a test, after which it is terminated and marked as failed. (**timeout**)

We track all our ideas for enhancement as [GitHub issues](https://github.com/kyma-incubator/octopus/issues) so you can easily refer to them for details.

As an open-source project, we always welcome external contributions. So if you only wish, you can help us in many ways:

- [Pick](https://github.com/kyma-incubator/octopus/issues) one of the existing issues and try to propose a solution for it in a pull request.
- [Add](https://github.com/kyma-incubator/octopus/issues/new/choose) your own issue with ideas for improving Octopus.
- [Star](https://github.com/kyma-incubator/octopus) Octopus to support us in our attempt to join the [Awesome Kubernetes](https://github.com/ramitsurana/awesome-kubernetes#testing) family of testing projects, where we believe Octopus could take pride of place.

> If you have any questions or want to find out more about us, contact us on the [`#octopus`](https://kyma-community.slack.com/archives/CGDKAE7J5) Slack channel or visit our [website](http://slack.kyma-project.io/).

Reviewer: it is not possible to link like this to the channel for new users, this is why http://slack.kyma-project.io/ (which is a link to the invitation) must be used here. The link that you provided will not work for new user and he will not know how to get into the slack

Author: (I also corrected the website link)

> But first of all, [test Octopus](https://github.com/kyma-incubator/octopus) on your own and see if it does the trick for you as well.

Reviewer 2: ```suggestion
But first of all, give [Octopus](https://github.com/kyma-incubator/octopus) a try and see if it does the trick for you as well.
```
just a preference, not forcing :)
